inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_b_fs_yx_fsv16_imad.cl

```diff
@@ -0,0 +1,191 @@
+        bool zero_y = input_offset_y >= INPUT0_SIZE_Y || input_offset_y < 0;
+        if(!zero_y)
+        {
+            for(uint i = 0; i < POOL_SIZE_X; i++)
+            {
+                int input_offset_x = offset_x + i;
+                bool zero = input_offset_x >= INPUT0_SIZE_X || input_offset_x < 0;
+                if(!zero)
+                {
+                    const uint input_idx = batch_and_feature_offset + input_offset_y*IN_Y_PITCH + input_offset_x*IN_X_PITCH;
+
+                    int4 int_data = vload4(0, (__global int*)(input + input_idx));
+                    IN_VEC16 ch16_data = AS_TYPE(IN_VEC16, int_data);
+                    for(uint k = 0; k < FEATURE_SLICE_SIZE; k++)
+                    {
+                        result[k] = FUNC_CALL(apply_pooling)(result[k], ch16_data[k]);
+                    }
+
+#ifdef DYNAMIC_KERNEL_DIVIDER
+                    num_elements++;
+#endif
+                }
+            }
+        }
+    }
+#ifdef DYNAMIC_WITH_PADDING_KERNEL_DIVIDER
+    const int hend = min(offset_y + POOL_SIZE_Y, INPUT0_SIZE_Y + PADDING_SIZE_Y);
+    const int wend = min(offset_x + POOL_SIZE_X, INPUT0_SIZE_X + PADDING_SIZE_X);
+    const uint num_elements = (hend - offset_y) * (wend - offset_x);
+#endif
+#else // !CHECK_BOUNDRY
+    uint input_idx = INPUT0_GET_INDEX(b, f, offset_y, offset_x);
+
+    for(uint j = 0; j < POOL_SIZE_Y; j++)
+    {
+        for(uint i = 0; i < POOL_SIZE_X; i++)
+        {
+            int4 int_data = vload4(0, (__global int*)(input + input_idx));
+            IN_VEC16 ch16_data = AS_TYPE(IN_VEC16, int_data);
+            for(uint k = 0; k < FEATURE_SLICE_SIZE; k++)
+            {
+                result[k] = FUNC_CALL(apply_pooling)(result[k], ch16_data[k]);
+            }
+
+            input_idx += IN_X_PITCH;
+        }
+        input_idx += (IN_Y_PITCH - POOL_SIZE_X*IN_X_PITCH);
+    }
+
+#if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+    const uint num_elements = POOL_SIZE_X*POOL_SIZE_Y;
+#endif
+#endif
+
+#if defined AVG_POOLING
+#if ENABLE_ROUND
+    int16 pool_result;
+    for(uint i = 0; i < FEATURE_SLICE_SIZE; i++) {
+    #if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+        pool_result[i] = convert_int(round(((float)result[i] / max(num_elements, (uint)1))));
+    #else
+        pool_result[i] = convert_int(round((float)result[i] / (int)(POOL_SIZE_Y * POOL_SIZE_X)));
+    #endif
+    }
+#else
+    float16 pool_result;
+    for(uint i = 0; i < FEATURE_SLICE_SIZE; i++) {
+    #if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+        pool_result[i] = (float)result[i] / max(num_elements, (uint)1);
+    #else
+        pool_result[i] = (float)result[i] / (int)(POOL_SIZE_Y * POOL_SIZE_X);
+    #endif
+    }
+#endif  // ENABLE_ROUND
+#else  // AVG_POOLING
+    int16 pool_result;
+    for (uint i = 0; i < FEATURE_SLICE_SIZE; ++i) {
```

Comment:
Why do you use pre-incrementation operator only in 2 last cycles? For getting better performance or for another reason?

Reply:
There isn't any specific reason. I just based this kernel on pooling b_fs_yx_fsv4 and bs_fs_yx_bsv16_fsv16 and that is why it looks like this.